paddle/fluid/operators/batch_norm_op.cu

```diff
@@ -158,23 +184,103 @@ class BatchNormKernel<platform::CUDADeviceContext, T>
       } else {
         double this_factor = 1. - momentum;
 
-        CUDNN_ENFORCE(platform::dynload::cudnnBatchNormalizationForwardTraining(
-            handle, mode_, CudnnDataType<T>::kOne(), CudnnDataType<T>::kZero(),
-            data_desc_, x->template data<T>(), data_desc_,
-            y->template mutable_data<T>(ctx.GetPlace()), bn_param_desc_,
-            scale->template data<BatchNormParamType<T>>(),
-            bias->template data<BatchNormParamType<T>>(), this_factor,
-            mean_out->template mutable_data<BatchNormParamType<T>>(
-                ctx.GetPlace()),
-            variance_out->template mutable_data<BatchNormParamType<T>>(
-                ctx.GetPlace()),
-            epsilon, saved_mean->template mutable_data<BatchNormParamType<T>>(
-                         ctx.GetPlace()),
-            saved_variance->template mutable_data<BatchNormParamType<T>>(
-                ctx.GetPlace())));
+        bool called = false;
+#if CUDNN_VERSION_MIN(7, 4, 1)
+        if (compute_format == "NHWC") {
+          called = true;
+          size_t workspace_size = 0;
+          size_t reserve_space_size = 0;
+          void *reserve_space_ptr = nullptr;
+          void *workspace_ptr = nullptr;
+          Tensor workspace_tensor;
+          // Create reserve space and workspace for batch norm.
+          // Create tensor for each batchnorm op, it will be used in the
+          // backward. Thus this tensor shouldn't be temp.
+          auto it = ctx.op().Inputs().find("X");
+          auto input_var_name = it->second[0];
+          Tensor *reserve_space_tensor =
```

Comment:
Here is an example: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/conv_cudnn_op.cu#L130